### PR TITLE
Fix leak in Digest.cc (bsc#1075978)

### PR DESCRIPTION
**Pull request by bzeller**

### CODE REVIEW

File: zypp/Digest.cc [hunk 1]
```diff
@@ -137,13 +143,8 @@ namespace zypp {
...
       if(initialized)
       {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
-        EVP_MD_CTX_cleanup(mdctx);
-#else
-        EVP_MD_CTX_free(mdctx);
-#endif
-    	initialized = false;
-    	finalized = false;
+        initialized = false;
```
Comment (mlandres) [inline zypp/Digest.cc]: I leave it up to you. Personally I'd eliminate `bool initialized` in `Digest::P`.

All tests for `initialized` can be replaced by tests for `mdctx`. Creation of the context in the `if(!initialized)` in `maybeInit` could use a local `EvpDataPtr` which is assigned to `mdctx` once all init steps succeeded (replaces `initialized = true;`). Here in `cleanup` the `initialized = false;` would be replaced by `mdctx.reset();` to invalidate the context.

File: zypp/Digest.cc [hunk 2]
```diff
@@ -67,7 +73,7 @@ namespace zypp {
     	P();
     	~P();
 
-        EVP_MD_CTX *mdctx;
+      EvpDataPtr mdctx;
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/Digest.cc]: Isn't it now aligned with surrounding code? The old definition was not, or do I miss something?

File: zypp/Digest.cc [hunk 3]
```diff
@@ -27,6 +27,12 @@
 #endif
 
 #include "zypp/Digest.h"
+#include "zypp/base/PtrTypes.h"
+#include "zypp/base/Logger.h"
+
+namespace {
+  typedef zypp::shared_ptr<EVP_MD_CTX> EvpDataPtr;
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/Digest.cc]: done

File: zypp/Digest.cc [hunk 4]
```diff
@@ -27,6 +27,12 @@
 #endif
 
 #include "zypp/Digest.h"
+#include "zypp/base/PtrTypes.h"
+#include "zypp/base/Logger.h"
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/Digest.cc]: Done

File: zypp/Digest.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlandres) [reply zypp/Digest.cc]: Maybe [github renders it wrong](https://github.com/bzeller/libzypp/blob/dd3a3b4c8a5c1c4c1286f59bbc8dd7d56f7d1526/zypp/Digest.cc#L76). If it's actually allinged it's fine.

[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/Digest.cc]: Done

[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/Digest.cc]: Seems github does not render with tabwidth==2 , the file has tabs and spaces mixed. Thats why it appears to be different.

File: zypp/Digest.cc (hunk 1 above)
Comment (bzeller) [reply zypp/Digest.cc]: Done

**Pull request merged**